### PR TITLE
refactor seo

### PR DESCRIPTION
**Pull request by iBakuman**
## The new features added in this PR are as follows

- SEO supports multi-level inheritance.
- Batch rendering is supported, and this functionality is implemented through the `BatchRender` method.
- The management page of SEO is implemented using Listing.
- Rename struct `Collection` to `Builder`.

## The following content has been removed:

- Removed `QorSettingInterface` interface and related methods.
- Removed `dbContextKey` and related methods.
- ...

### CODE REVIEW

File: seo/builder.go [hunk 1]
```diff
@@ -0,0 +1,375 @@
...
+		b.locales = locales
+	}
+}
+
+func NewBuilder(db *gorm.DB, ops ...Option) *Builder {
+	globalSEO := &SEO{name: defaultGlobalSEOName}
+	globalSEO.RegisterSettingVariables(struct {
+		SiteName string
+	}{})
+	globalSEO.RegisterPropFuncForOG(&PropFunc{
```
Comment (sunfmin) [inline seo/builder.go:58]: globalSEO.RegisterOpenGraphProperty("og:url", 
                 func(_ interface{}, _ *Setting, req *http.Request) string {
			return req.URL.String()
		}
	)

Comment (raven-chen) [reply seo/builder.go:58]: Decided to remove this default function and `PropFunc`. but keep `RegisterMetaProperty` function for explicit invoke

File: seo/builder.go [hunk 2]
```diff
@@ -0,0 +1,375 @@
...
+
+func WithLocales(locales ...string) Option {
+	return func(b *Builder) {
+		b.locales = locales
+	}
+}
+
+func NewBuilder(db *gorm.DB, ops ...Option) *Builder {
+	globalSEO := &SEO{name: defaultGlobalSEOName}
+	globalSEO.RegisterSettingVariables(struct {
```
Comment (raven-chen) [inline seo/builder.go:55]: @iBakuman  please update this to `RegisterSettingVariables(name...)`

File: seo/builder.go [hunk 3]
```diff
@@ -0,0 +1,383 @@
...
+	"gorm.io/gorm/clause"
+)
+
+const (
+	defaultGlobalSEOName = "Global SEO"
+	defaultLocale        = "en"
+)
+
+type (
+	contextVariablesFunc func(interface{}, *Setting, *http.Request) string
```
Comment (xuxinx) [inline seo/builder.go:26]: make it public?

Comment (iBakuman) [reply seo/builder.go:26]: Okay, making it public can make it easier for users to create context function.

File: seo/builder.go [hunk 4]
```diff
@@ -0,0 +1,373 @@
...
+	}
+	localeFinalSeoSetting := seo.getLocaleFinalQorSEOSetting(locale, b.db)
+	return b.render(obj, localeFinalSeoSetting, seo, req)
+}
+
+// BatchRender rendering multiple SEOs at once.
+// It is the responsibility of the caller to ensure that every element in objs
+// is of the same type, as it is performance-intensive to check whether each element
+// in `objs` if of the same type through reflection.
+func (b *Builder) BatchRender(objs []interface{}, req *http.Request) []h.HTMLComponent {
```
Comment (xuxinx) [inline seo/builder.go:235]: please change the type of `objs` to `interface{}` to accept any type of slice

Comment (iBakuman) [reply seo/builder.go]: It has been changed successfully.

**Review by xuxinx — APPROVED** [review]

**Pull request merged**